### ISSUE TITLE
Adding the afterRender() lifecycle method

### ISSUE THREAD

**Issue by jeswin** [forgojs/forgo#20]
Edit: The first proposal suggested getting rid of (/renaming) mount(). Retracted.

Currently, we have two lifecycle methods - mount and unmount.

unmount() is very clear, it happens when the component goes away (and the underlying node is removed from the DOM tree). It's a good place to write tear-down logic.

mount() - is a little tricky. Currently it gets called the first time a component is mounted on a node, and does not get called when it gets reattached to another node. Re-attach happens, for example when a CustomComponent renders a DIV tag in the first pass, and a P tag in the second pass. It is useful as a function that gets called only once for a component, and never afterwards. render() on the other hand, gets called multiple times.

For application developers, mount was never necessary. Whatever you could write in mount() could already be written inside the component constructor or within a component's render() method.

So, for people who want to detect when the DOM node changes, I propose adding a new method called afterRender() which fires just after the rendered components are attached to actual nodes. At that point, args.element would point to the updated DOM node.

@spiffytech @jacob-ebey Any thoughts?

**Comment by spiffytech** [comment]
It sounds like this new lifecycle method would run when `render()` response causes [renderDOMElement to change the element type](https://github.com/forgojs/forgo/blob/main/src/index.ts#L282)? And this method would only be called when the DOM element *changed*, and not on first render (that's `mount()`'s job)?

I definitely like giving developers a way to hook into that happening. That's important for many use cases where a dev wants to pass a raw DOM reference to a vanilla JS library such as D3.

Off the cuff, I think exposing this as a lifecycle method makes a lot of sense. The application developer would be in full control over any code that would cause this lifecycle method to fire, but I like letting the framework detect this for them rather than making them vigilantly track the situation by hand.

A dev might want `afterRender()` to respond to something they computed inside `render()`, whatever logic or data caused `render()` to change element types, but that should be derivable from ctor variables and from props passed to both lifecycle methods, so the common answer there would be lifting the logic into a reusable function inside the component ctor, If that's inappropriate for a given situation (expensive to compute, maybe?), using a ctor variable to shuttle data between lifecycle methods works, but it doesn't feel great.

I'd like to find a more elegant solution for passing whatever computed values led `render()` to change the element type over to `afterRender()`, but at first blush I think cases where both methods need a value and can't reasonably compute it from a ctor-level function would be uncommon, and we could probably come up with a good answer in a future PR without breaking the API chosen in this issue.

**Comment by jeswin** [comment]
Great points. Like you said, the application developer will be able to know if their render() is going to cause the node to change. But the problem is that currently they don't have a way to get a callback when the actual node-attachment happens.

Sorry, I am not entirely sure if I fully followed your 2nd-last paragraph - could you explain with a tiny example? 

I think if we're going to call it afterRender(), it should run after render() irrespective of whether it causes a node change or not. So one way (for a library developer) to check whether the node changed in afterRender() would be to store args.element as follows:

```jsx
function Component() {
  let prevElement;
  afterRender(props, args) {
    if (prevElement !== args.element) {
      console.log("Node changed");
    }
    prevElement = args.element;
  }
}
```

Another option is to have a method called attachNode() instead; which will only get called when the underlying node changes. But since afterRender() looks more versatile and can already achieve the same result (albeit with a couple of extra lines of code), perhaps it's better to just have afterRender()?

**Comment by jeswin** [comment]
@spiffytech And as always, your feedback and thoughts have been very helpful. Thanks.

**Comment by spiffytech** [comment]
> one way (for a library developer) to check whether the node changed in afterRender()

Tracking the previous node seems like bookkeeping work Forgo could handle itself without much trouble, which would reduce developer mistakes and keep application code simple. What about something like this?

```
function Component() {
  afterRender(props, args) {
    if (args.prevNode !== args.element.node) {
      console.log("Node changed");
    }
  }
}
```

or 

```
function Component() {
  afterRender(props, args, previousNode) {
    if (previousNode !== args.element.node) {
      console.log("Node changed");
    }
  }
}
```
> Another option is to have a method called attachNode()

I like `afterRender()`. As you said, if `afterRender()` can do it anyway, it seems better to have a small number of lifecycle methods that elegantly compose into complex behavior.

> Sorry, I am not entirely sure if I fully followed your 2nd-last paragraph - could you explain with a tiny example?

This is one of those things where my gut as an experienced dev is telling me something, and my rational mind is trying to play catch up and think through what my gut is telling me, and whether my gut is actually right. I tried contriving a concrete code scenario where what I was describing made any sense, but the more I think through it, the more I see that Forgo already has superior ways to solve every situation I think of. So, never mind :) Consider that part a non-issue for now.

**Comment by jeswin** [comment]
args.previousNode (or previousElement maybe) is a nice solution - I like it. 

The second option you presented adds a parameter, and that tempts us to do the same thing is future - which perhaps we should avoid (the order etc might become harder to remember). And in any case, your first solution will work well with destructuring - so we're not going to need additional keystrokes.

So that's our solution then, thanks. I'll land it along with fragment support, so it's going to take a little longer.

**Comment by jeswin** [comment]
Implemented in v0.0.36 along with Fragment support. I'll close this for now - and track bugs separately.
